### PR TITLE
remove multi dev parsing from xbutil and xbmgmt sub commands/options

### PR DESCRIPTION
**Pull request by AShivangi**
Related CRs and Story:
- [CR-1109413](https://jira.xilinx.com/browse/CR-1109413) ASTeR TS 18.02 - u55c/n+X3+u200 - long xclbin program times when executing in parallel on multiple cards
- [CR-1128477](https://jira.xilinx.com/browse/CR-1128477) ASTeR 3.01/02/03 - vck5000 - after a reboot xbutil examine may error in a Multi-card setup with the APU package installed
- [VITIS-4023](https://jira.xilinx.com/browse/VITIS-4023) - Improve the behavior for starting up utilities (xbutil/xbmgmt)
#### Problem solved by the commit
1st iteration of: [VITIS-4023](https://jira.xilinx.com/browse/VITIS-4023) - Improve the behavior for starting up utilities (xbutil/xbmgmt)

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
xbutil/xbmgmt no longer support multi-device operations. Cleaning up the code to only open/get the device that has been requested

#### Risks (if any) associated the changes in the commit
None

#### What has been tested and how, request additional testing if necessary
Locally tested:
1. No device specified
```
$ xbutil reset 
ERROR: Please specify a device using --device option
 Available devices:
  [0000:b3:00.1] : xilinx_u200_gen3x16_xdma_base_2
  [0000:65:00.1] : xilinx_vck5000_gen4x8_xdma_base_2
  [0000:18:00.1] : xilinx_u30_gen3x4_base_2
  [0000:17:00.1] : xilinx_u30_gen3x4_base_2
```

2. 1 valid device
```
$ xbutil reset -d b3:00
Performing 'HOT Reset' on '0000:b3:00.1'
Are you sure you wish to proceed? [Y/n]: 
Successfully reset Device[0000:b3:00.1]

```

3. single device system
```
$ xbutil reset -d
Performing 'HOT Reset' on '0000:65:00.1'
Are you sure you wish to proceed? [Y/n]: y
Successfully reset Device[0000:65:00.1]
```

4. 1 invalid device
```
$ xbutil reset -d b4:00
ERROR: Specified device BDF 'b4:00' not found
 Available devices:
  [0000:b3:00.1] : xilinx_u200_gen3x16_xdma_base_2
  [0000:65:00.1] : xilinx_vck5000_gen4x8_xdma_base_2
  [0000:18:00.1] : xilinx_u30_gen3x4_base_2
  [0000:17:00.1] : xilinx_u30_gen3x4_base_2
```
#### Documentation impact (if any)
None

### CODE REVIEW

**Comment by dayeh-xilinx** [comment]
retest this please. failed due to an infra issue (cfgmgr permission)

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/common/system.h [hunk 1]
```diff
@@ -67,6 +67,11 @@ class system
   virtual device::id_type
   get_device_id(const std::string& str) const;
 
+  /**
+   */
+  virtual std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user = true) const = 0;//{ return std::make_tuple(0, 0, 0, 0); }
```
Comment (stsoe) [inline src/runtime_src/core/common/system.h:73]: Let's remove the trailing comment and instead expand the leading comment.
I am thinking maybe you do want the default implementation as indicated in the trailing comment and not have to change the emulation shims that do not support bdf.

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/system_swemu.cxx [hunk 2]
```diff
@@ -57,6 +57,13 @@ get_total_devices(bool is_user) const
   return {0,0};
 }
 
+std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+system::
+get_bdf_info(device::id_type id, bool is_user) const
+{
+  return std::make_tuple(0, 0, 0, 0);
+}
+
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/system_swemu.cxx:66]: Can we have a default implementation in base class?

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/system_swemu.h [hunk 3]
```diff
@@ -29,6 +29,9 @@ class system : public system_edge
   std::pair<device::id_type, device::id_type>
   get_total_devices(bool is_user) const;
 
+  std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user = true) const;
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/system_swemu.h:33]: Not needed if we just use a default base class implementation.

File: src/runtime_src/core/edge/user/system_linux.cpp [hunk 4]
```diff
@@ -171,6 +171,13 @@ get_total_devices(bool is_user) const
   return std::make_pair(num, num);
 }
 
+std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+system_linux::
+get_bdf_info(device::id_type id, bool is_user) const
+{
+  return std::make_tuple(0,0,0,0);
+}
+
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/system_linux.cpp:180]: ditto

File: src/runtime_src/core/edge/user/system_linux.h [hunk 5]
```diff
@@ -33,6 +33,9 @@ class system_linux : public system_edge
   std::pair<device::id_type, device::id_type>
   get_total_devices(bool is_user) const;
 
+  std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user = true) const;
+
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/system_linux.h:38]: ditto

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/system_swemu.cxx [hunk 6]
```diff
@@ -58,6 +58,13 @@ get_total_devices(bool is_user) const
   return {nd,nd};
 }
 
+std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+system::
+get_bdf_info(device::id_type id, bool is_user) const
+{
+  return std::make_tuple(0, 0, id, (is_user ? 1 : 0));
+}
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/system_swemu.cxx:66]: Really?  Do we care to have this special implementation?  There is no bdf in swemu.

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/system_swemu.h [hunk 7]
```diff
@@ -29,6 +29,9 @@ class system : public system_pcie
   std::pair<device::id_type, device::id_type>
   get_total_devices(bool is_user) const;
 
+  std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user) const;
+
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/system_swemu.h:34]: Do we need?

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/system_hwemu.cxx [hunk 8]
```diff
@@ -58,6 +58,13 @@ get_total_devices(bool is_user) const
   return {nd,nd};
 }
 
+std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+system::
+get_bdf_info(device::id_type id, bool is_user) const
+{
+  return std::make_tuple(0, 0, id, (is_user ? 1 : 0));
+}
+
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/system_hwemu.cxx:67]: same comment as ealier

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/system_hwemu.h [hunk 9]
```diff
@@ -29,6 +29,9 @@ class system : public system_pcie
   std::pair<device::id_type, device::id_type>
   get_total_devices(bool is_user) const;
 
+  std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user) const;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/system_hwemu.h:33]: ditto

File: src/runtime_src/core/pcie/noop/system_noop.cpp [hunk 10]
```diff
@@ -55,6 +55,13 @@ get_total_devices(bool is_user) const
   return {0,0};
 }
 
+std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+system::
+get_bdf_info(device::id_type id, bool is_user) const
+{
+  return std::make_tuple(0, 0, 0, 0);
+}
+
```
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/system_noop.cpp:64]: Not needed

File: src/runtime_src/core/pcie/noop/system_noop.h [hunk 11]
```diff
@@ -29,6 +29,9 @@ class system : public system_pcie
   std::pair<device::id_type, device::id_type>
   get_total_devices(bool is_user) const;
 
+  std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user) const;
+
```
Comment (stsoe) [inline src/runtime_src/core/pcie/noop/system_noop.h:34]: Not needed.

File: src/runtime_src/core/pcie/windows/mcdm/system_mcdm.cpp [hunk 12]
```diff
@@ -45,6 +45,13 @@ get_total_devices(bool) const
   return {count, count};
 }
 
+std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+system_mcdm::
+get_bdf_info(device::id_type id, bool is_user) const
+{
+  return std::make_tuple(0, 0, id, (is_user ? 1 : 0));
+}
+
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/mcdm/system_mcdm.cpp:54]: Not needed.  In fact, this mcdm directory should be removed.

File: src/runtime_src/core/pcie/windows/mcdm/system_mcdm.h [hunk 13]
```diff
@@ -13,6 +13,9 @@ class system_mcdm : public system_pcie
   std::pair<device::id_type, device::id_type>
   get_total_devices(bool is_user) const;
 
+  std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user) const;
+
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/mcdm/system_mcdm.h:18]: Not needed

**Review by stsoe — CHANGES_REQUESTED** [review]
I think a default impl of get_bdf_info is warranted.

File: src/runtime_src/core/common/system.h (hunk 1 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/system.h:73]: I was not sure what was the best choice as total devices is implemented everywhere but I am glad that is the right way to do it!

File: src/runtime_src/core/common/system.h [hunk 14]
```diff
@@ -67,6 +67,11 @@ class system
   virtual device::id_type
   get_device_id(const std::string& str) const;
 
+  /**
+   */
+  virtual std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type id, bool is_user = true) const { return std::make_tuple(0, 0, 0, 0); }
```
Comment (stsoe) [inline src/runtime_src/core/common/system.h:73]:
```suggestion
  get_bdf_info(device::id_type id, bool is_user = true) const 
  { 
     return std::make_tuple(0, 0, 0, 0); 
   }
```

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/system.h [hunk 15]
```diff
@@ -67,6 +67,14 @@ class system
   virtual device::id_type
   get_device_id(const std::string& str) const;
 
+  /**
+   */
+  virtual std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info(device::id_type, bool) const
+  {
+    return std::make_tuple(uint16_t(0), uint16_t(0), uint16_t(0), uint16_t(0));
```
Comment (stsoe) [inline src/runtime_src/core/common/system.h:75]:
```suggestion
    return {0, 0, 0, 0};
```

If that doesn't work for some reason, then use 
`std::make_tuple<uint16_t, uint16_t, uint16_t, uint16_t>(0, 0, 0, 0)`

**Review by stsoe — CHANGES_REQUESTED** [review]
I checked the infrastructure changes, not the tools changes.
It looks good, and simplifies code.
One minor request.

Plus do update / change preamble and copyright to include
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2022 Advanced Micro Devices, Inc. All rights reserved.
It seems like AMD's Palamida scans require this.  Do not remove any existing copyrights.

File: src/runtime_src/core/common/system.h (hunk 15 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/system.h:75]: Unfortunately, neither of those two options will compile in the windows build.

File: src/runtime_src/core/pcie/windows/alveo/system_windows.cpp [hunk 16]
```diff
@@ -13,19 +12,25 @@
 #include "device_windows.h"
```
Comment (rozumx) [inline src/runtime_src/core/pcie/windows/alveo/system_windows.cpp:12]: Include list should be sorted.

File: src/runtime_src/core/tools/common/XBHelpMenus.cpp [hunk 17]
```diff
@@ -221,85 +209,83 @@ XBUtilities::produce_reports( xrt_core::device_collection devices,
...
+        platform = xrt_core::device_query<xrt_core::query::rom_vbnv>(device);
       }
+    } 
+    catch (const xrt_core::query::exception&) {
+      // proceed even if the platform name is not available
+      platform = "<not defined>";
+    }
+    const std::string dev_desc = (boost::format("[%s] : %s\n") % ptDevice.get<std::string>("device_id") % platform).str();
+    consoleStream << std::endl;
+    consoleStream << std::string(dev_desc.length(), '-') << std::endl;
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/XBHelpMenus.cpp:241]: Add a comment to let the user that this code is bounding the device description with minus '-'

File: src/runtime_src/core/tools/common/XBHelpMenus.cpp [hunk 18]
```diff
@@ -221,85 +209,83 @@ XBUtilities::produce_reports( xrt_core::device_collection devices,
...
+    }
+    catch(const xrt_core::query::exception&) { 
+      is_recovery = false;
+    }
+
+    // Process the tests that require a device
+    // If the device is either of the following, most tests cannot be completed fully:
+    // 1. Is in factory mode and is not in recovery mode
+    // 2. Is not ready and is not in recovery mode
+    if ((is_mfg || !is_ready) && !is_recovery) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/XBHelpMenus.cpp:258]: Curly braces are not needed.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 19]
```diff
@@ -751,7 +738,7 @@ SubCmdProgram::execute(const SubCmdOptions& _options) const
   }
 
   // -- Retrieve and parse the subcommand options -----------------------------
-  std::vector<std::string> device;
+  std::string device_str;
   std::string plp = "";
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:742]: The string class by default initialized the string to an empty string as such there isn't a need for this extra assignment.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 20]
```diff
@@ -938,17 +891,14 @@ SubCmdProgram::execute(const SubCmdOptions& _options) const
...
-        xrt_core::error(boost::str(boost::format("%d is an invalid index") % dev->get_device_id()));
-        continue;
-      }
-      std::cout << boost::format("%-8s : %s %s %s \n") % "INFO" % "Resetting device ["
-        % flasher.sGetDBDF() % "] back to factory mode.";
-      flasher_list.push_back(flasher);
-    }
+    //collect information of all the devices that will be reset
+    Flasher flasher(device->get_device_id());
+    if(!flasher.isValid())
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:896]:
```suggestion
    if (!flasher.isValid())
```

**Review by rozumx — APPROVED** [review]
Minor change requests.

**Review by chvamshi-xilinx — APPROVED** [review]

**Pull request merged**